docs/build-provenance-for-all-registries.md

```diff
@@ -0,0 +1,104 @@
+
+Now you’re ready to sign the package! You’ll create an ephemeral key pair, and then send to Fulcio the OIDC token, your ephemeral public key, the OIDC subject signed with your private key (the “challenge value”), and your build artifact signed with your private key. See [sigstore-js' sign.ts](https://github.com/sigstore/sigstore-js/blob/main/packages/client/src/sign.ts#L110) for a reference implementation.
+
+Fulcio will return to you a X.509 signing certificate, with the build provenance properties (including your public key) as OIDs in that certificate. You can now throw away your ephemeral private key.
+
+But we aren’t quite ready to send things to the registry - first you’ll sign your build attestation document (with the Fulcio signing certificate), publish it to Sigstore’s Rekor to get a transparency log entry, and use that to create your Sigstore signature bundle with your signatures over your build artifact, the build attestation document, your transparency log entry, and a timestamp.
+
+You might be wondering how much of this Sigstore interaction you need to implement yourself. Ideally there would be a mature Sigstore library in a language supported by your ecosystem tooling. There are currently libraries for [JavaScript](https://github.com/sigstore/sigstore-js), [Go](https://github.com/sigstore/sigstore-go), [Python](https://github.com/sigstore/sigstore-python), as well libraries under development for [Rust](https://github.com/sigstore/sigstore-rs), [Ruby](https://github.com/sigstore/ruby-sigstore), and [Java](https://github.com/sigstore/sigstore-java). See [sigstore-conformance](https://github.com/sigstore/sigstore-conformance) to ensure client libraries are behaving as expected.
+
+
+# Extending the registry to support build provenance
+
+Now we’re ready to publish the package to the registry, but we need a way to authorize the request. Instead of requiring maintainers to store their password or an unscoped API token in the cloud CI/CD provider, we recommend allowing maintainers to set up a trust relationship between the cloud CI/CD system and your registry via OIDC. For an example of this, see <https://docs.pypi.org/trusted-publishers/>. In fact, registries can implement OIDC trust relationships to authorize publish requests without needing any support from Sigstore client libraries. So this is an excellent scoped project to start with, which is exactly what PyPI did.
+
+Finally, we submit the package and the signed build attestations to the registry. Our work is done… as the CLI tooling! Now we need to verify the provenance on the registry side. See above for a discussion on what Sigstore tooling is available in what programming languages.
```

Comment:
"we" is doing a lot of work here, which is a bit clever, but may not be entirely clear.  I'd suggest making the subject more explicit:

> Once the packaging CLI has submitted the package and the signed build attestations to the registry, the registy needs to verify these artifacts.  This should be done before the artifact is published (made visible) by the registry.  The sigstore libraries mentioned above may be helpful in this process.

Reply:
The intent was to avoid explicitly referring to the CLI tooling, as above we have that as a recommendation and not a requirement:

> We recommend building this functionality into a CLI tool. Most ecosystems already have some CLI tool involved in the package publishing process, and this allows the implementation to be shared across CI/CD providers.

... but thinking about it more, maybe that's silly, as every ecosystem we've talked to has CLI tooling for interacting with their package registry. Regardless, I reworded this section.